File: Medium/57. Insert Interval.ts
function insert(intervals: number[][], newInterval: number[]): number[][] {
    let n = intervals.length,
    i = 0,
    res = [];

// Case 1: No overlapping before merging intervals
while (i < n && intervals[i][1] < newInterval[0]) {
    res.push(intervals[i]);
    i++;
}

// Case 2: Overlapping and merging intervals
while (i < n && newInterval[1] >= intervals[i][0]) {
    newInterval[0] = Math.min(newInterval[0], intervals[i][0]);
    newInterval[1] = Math.max(newInterval[1], intervals[i][1]);
    i++;
}
res.push(newInterval);

// Case 3: No overlapping after merging newInterval
while (i < n) {
    res.push(intervals[i]);
    i++;
}

return res;
};

File: python/57. Insert Interval.py
class Solution(object):
    def insert(self, intervals, newInterval):
        n = len(intervals)
        i = 0
        res = []

        # Case 1: intervals not overlapping and before newInterval
        while i < n and intervals[i][1] < newInterval[0]:
            res.append(intervals[i])
            i += 1

        # Case 2: intervals overlapping with newInterval
        while i < n and newInterval[1] >= intervals[i][0]:
            newInterval[0] = min(newInterval[0], intervals[i][0])
            newInterval[1] = max(newInterval[1], intervals[i][1])
            i += 1

        # After merging all overlaps, add the merged newInterval
        res.append(newInterval)

        # Case 3: remaining intervals (those come after newInterval)
        while i < n:
            res.append(intervals[i])
            i += 1

        return res

        """
        :type intervals: List[List[int]]
        :type newInterval: List[int]
        :rtype: List[List[int]]
        """